File: runtimes/switch/genai_span.mjs
// runtimes/switch/genai_span.mjs -- GenAI semantic-convention spans, no SDK.
//
// ADR 0058 / GAP 74: OTEL is the vocabulary, not the floor, and not a
// dependency. This file emits gen_ai.* attributes as JSON. Local append
// happens first. OTLP is optional (OTEL_EXPORTER_OTLP_TRACES_ENDPOINT)
// and must not fail the request. Logfire is not imported. Prompts,
// messages, and credentials never become attributes.

import { appendFileSync } from 'node:fs';
import { randomBytes } from 'node:crypto';

export const SCOPE_NAME = 'switch/genai';
export const SCOPE_VERSION = '1';

const FORBIDDEN = [
  'authorization', 'cookie', 'set-cookie', 'session_token', 'api_key',
  'password', 'email', 'token', 'prompt', 'messages', 'content',
  'completion', 'gen_ai.input.messages', 'gen_ai.output.messages',
  'gen_ai.prompt', 'gen_ai.completion',
];

export function hexId(bytes) {
  return randomBytes(bytes).toString('hex');
}

export function parseTraceparent(header) {
  const s = String(header || '').trim();
  const parts = s.split('-');
  if (parts.length >= 4 && /^[0-9a-f]{32}$/i.test(parts[1]) && /^[0-9a-f]{16}$/i.test(parts[2])) {
    return { traceId: parts[1].toLowerCase(), parentSpanId: parts[2].toLowerCase() };
  }
  return { traceId: hexId(16), parentSpanId: '' };
}

function dropForbidden(attrs) {
  const out = {};
  for (const [k, v] of Object.entries(attrs || {})) {
    const key = String(k);
    if (FORBIDDEN.includes(key.toLowerCase()) || FORBIDDEN.includes(key)) continue;
    if (v == null || v === '') continue;
    out[key] = v;
  }
  return out;
}

export function chatSpan({ provider, model, traceparent, status, reason, usage, durationMs }) {
  const ids = parseTraceparent(traceparent);
  const name = `chat ${model || provider || 'unknown'}`;
  const attributes = dropForbidden({
    'gen_ai.operation.name': 'chat',
    'gen_ai.provider.name': provider || undefined,
    'gen_ai.request.model': model || undefined,
    'gen_ai.usage.input_tokens': usage && usage.input,
    'gen_ai.usage.output_tokens': usage && usage.output,
    'error.type': status === 'ERROR' ? (reason || 'error') : undefined,
  });
  return {
    name,
    kind: 'CLIENT',
    traceId: ids.traceId,
    spanId: hexId(8),
    parentSpanId: ids.parentSpanId || undefined,
    status: { code: status === 'ERROR' ? 'ERROR' : 'UNSET' },
    durationMs: Number(durationMs) || 0,
    attributes,
    'otel.scope.name': SCOPE_NAME,
    'otel.scope.version': SCOPE_VERSION,
  };
}

export function usageFromBody(text) {
  if (!text || typeof text !== 'string') return null;
  try {
    const j = JSON.parse(text);
    const u = j && j.usage;
    if (!u || typeof u !== 'object') return null;
    const input = u.input_tokens ?? u.prompt_tokens;
    const output = u.output_tokens ?? u.completion_tokens;
    if (input == null && output == null) return null;
    return { input, output };
  } catch {
    return null;
  }
}

function writeLocal(span) {
  const line = JSON.stringify({ gen_ai_span: span });
  const path = String(process.env.GENAI_SPAN_LOG || '').trim();
  if (path) {
    appendFileSync(path, line + '\n');
    return;
  }
  console.log(line);
}

async function exportOtlp(span) {
  const url = String(process.env.OTEL_EXPORTER_OTLP_TRACES_ENDPOINT || '').trim();
  if (!url) return;
  const attrs = Object.entries(span.attributes || {}).map(([key, value]) => {
    if (typeof value === 'number') return { key, value: { intValue: String(Math.trunc(value)) } };
    return { key, value: { stringValue: String(value) } };
  });
  const body = {
    resourceSpans: [{
      resource: { attributes: [{ key: 'service.name', value: { stringValue: 'switch' } }] },
      scopeSpans: [{
        scope: { name: SCOPE_NAME, version: SCOPE_VERSION },
        spans: [{
          traceId: span.traceId,
          spanId: span.spanId,
          parentSpanId: span.parentSpanId || '',
          name: span.name,
          kind: 3,
          startTimeUnixNano: '0',
          endTimeUnixNano: String(Math.max(0, Math.trunc((span.durationMs || 0) * 1e6))),
          status: { code: span.status && span.status.code === 'ERROR' ? 2 : 0 },
          attributes: attrs,
        }],
      }],
    }],
  };
  await fetch(url, {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(body),
  });
}

export function emitChat(args) {
  try {
    const span = chatSpan(args);
    writeLocal(span);
    exportOtlp(span).catch(() => {});
    return span;
  } catch {
    return null;
  }
}

File: gems/rails-cpcp/lib/rails_cpcp/genai_span.rb
# frozen_string_literal: true
require "json"
require "securerandom"

module RailsCpcp
  # GenAI semantic-convention spans at /_cpcp. No OTEL SDK (GAP 74).
  # Local append is the floor; OTLP is optional and never raises.
  # Logfire is not a dependency. Prompts and credentials are not attributes.
  module GenaiSpan
    module_function

    SCOPE_NAME = "rails-cpcp/genai"
    SCOPE_VERSION = "1"
    ENV_LOG = "CPCP_GENAI_SPAN_LOG"
    FORBIDDEN = %w[
      authorization cookie set-cookie session_token api_key password email token
      prompt messages content completion gen_ai.input.messages gen_ai.output.messages
      gen_ai.prompt gen_ai.completion
    ].freeze

    def emit(operation:, suffix:, kind:, traceparent: nil, ok: true, reason: nil, attributes: {}, duration_ms: 0)
      span = build(
        operation: operation,
        suffix: suffix,
        kind: kind,
        traceparent: traceparent,
        ok: ok,
        reason: reason,
        attributes: attributes,
        duration_ms: duration_ms
      )
      write_local(span)
      export_otlp(span)
      span
    rescue StandardError
      nil
    end

    def around(request:, traceparent: nil)
      t0 = monotonic_ms
      env = nil
      begin
        env = yield
      ensure
        method = (request && request["method"]).to_s
        params = (request && request["params"]) || {}
        opid = (request && (request["operationId"] || params["operationId"])).to_s
        failed = env.is_a?(Hash) && env["ok"] == false
        emit(
          operation: "invoke_agent",
          suffix: method,
          kind: "SERVER",
          traceparent: traceparent,
          ok: !failed,
          reason: failed ? env.dig("error", "reason") : nil,
          attributes: {
            "rpc.method" => method,
            "cpcp.operation_id" => opid,
            "gen_ai.agent.name" => "cpcp"
          },
          duration_ms: monotonic_ms - t0
        )
      end
      env
    end

    def build(operation:, suffix:, kind:, traceparent:, ok:, reason:, attributes:, duration_ms:)
      ids = parse_traceparent(traceparent)
      attrs = compact_attrs(
        { "gen_ai.operation.name" => operation.to_s }.merge(attributes || {})
      )
      attrs["error.type"] = reason.to_s if ok == false && reason && !reason.to_s.empty?
      {
        "name" => "#{operation} #{suffix}".strip,
        "kind" => kind.to_s,
        "traceId" => ids[:trace_id],
        "spanId" => SecureRandom.hex(8),
        "parentSpanId" => ids[:parent_span_id],
        "status" => { "code" => ok == false ? "ERROR" : "UNSET" },
        "durationMs" => duration_ms.to_f,
        "attributes" => attrs,
        "otel.scope.name" => SCOPE_NAME,
        "otel.scope.version" => SCOPE_VERSION
      }.compact
    end

    def parse_traceparent(header)
      parts = header.to_s.strip.split("-")
      if parts.length >= 4 && parts[1].match?(/\A[0-9a-f]{32}\z/i) && parts[2].match?(/\A[0-9a-f]{16}\z/i)
        { trace_id: parts[1].downcase, parent_span_id: parts[2].downcase }
      else
        { trace_id: SecureRandom.hex(16), parent_span_id: nil }
      end
    end

    def compact_attrs(attrs)
      out = {}
      attrs.each do |key, value|
        k = key.to_s
        next if FORBIDDEN.include?(k) || FORBIDDEN.include?(k.downcase)
        next if value.nil? || value.to_s.empty?

        out[k] = value
      end
      out
    end

    def log_path
      explicit = ENV[ENV_LOG].to_s
      return explicit unless explicit.empty?

      File.join(File.dirname(RefusalLog.log_path), "cpcp_genai_spans.jsonl")
    end

    def write_local(span)
      line = JSON.generate("gen_ai_span" => span)
      File.open(log_path, "a") { |io| io.puts(line) }
    end

    def export_otlp(span)
      url = ENV["OTEL_EXPORTER_OTLP_TRACES_ENDPOINT"].to_s.strip
      return if url.empty?

      require "net/http"
      require "uri"
      uri = URI.parse(url)
      return unless uri.is_a?(URI::HTTP)

      body = {
        "resourceSpans" => [{
          "resource" => { "attributes" => [{ "key" => "service.name", "value" => { "stringValue" => "cpcp" } }] },
          "scopeSpans" => [{
            "scope" => { "name" => SCOPE_NAME, "version" => SCOPE_VERSION },
            "spans" => [{
              "traceId" => span["traceId"],
              "spanId" => span["spanId"],
              "parentSpanId" => span["parentSpanId"].to_s,
              "name" => span["name"],
              "kind" => 2,
              "attributes" => (span["attributes"] || {}).map { |k, v| otlp_attr(k, v) }
            }]
          }]
        }]
      }
      http = Net::HTTP.new(uri.host, uri.port)
      http.open_timeout = 1
      http.read_timeout = 1
      req = Net::HTTP::Post.new(uri.request_uri)
      req["content-type"] = "application/json"
      req.body = JSON.generate(body)
      http.request(req)
    rescue StandardError
      nil
    end

    def otlp_attr(key, value)
      if value.is_a?(Numeric)
        { "key" => key, "value" => { "intValue" => value.to_i.to_s } }
      else
        { "key" => key, "value" => { "stringValue" => value.to_s } }
      end
    end

    def monotonic_ms
      Process.clock_gettime(Process::CLOCK_MONOTONIC, :millisecond)
    end
  end
end
